### PR TITLE
JSON schema validation and WOF document property normalisation

### PR DESCRIPTION
**Pull request by vicchi**
See https://github.com/whosonfirst/whosonfirst-json-schema/issues/1 for details of this PR in context with other changes/pull requests.

### CODE REVIEW

File: scripts/wof-fix-props [hunk 1]
```diff
@@ -0,0 +1,561 @@
...
+        f = geojson.load(fh)
+        fh.close()
+
+        fixed = False
+
+        # Fixup gn: properties
+        if 'gn:pop' in f['properties']:
+            if not isinstance(f['properties']['gn:pop'], basestring):
+                f['properties']['gn:pop'] = str(f['properties']['gn:pop'])
+                fixed = True
```
Comment (nvkelso) [inline scripts/wof-fix-props]: Int, no?

File: scripts/wof-fix-props [hunk 2]
```diff
@@ -0,0 +1,561 @@
...
+
+        if 'ne:geou_dif' in f['properties']:
+            if not isinstance(f['properties']['ne:geou_dif'], int):
+                f['properties']['ne:geou_dif'] = int(f['properties']['ne:geou_dif'])
+                fixed = True
+
+        if 'ne:iso_n3' in f['properties']:
+            if not isinstance(f['properties']['ne:iso_n3'], basestring):
+                f['properties']['ne:iso_n3'] = str(f['properties']['ne:iso_n3'])
```
Comment (nvkelso) [inline scripts/wof-fix-props]: Int

File: scripts/wof-fix-props [hunk 3]
```diff
@@ -0,0 +1,561 @@
...
+                fixed = True
+
+        if 'qs:loc_lc' in f['properties']:
+            if not isinstance(f['properties']['qs:loc_lc'], basestring):
+                f['properties']['qs:loc_lc'] = str(f['properties']['qs:loc_lc'])
+                fixed = True
+
+        if 'qs:photos' in f['properties']:
+            if not isinstance(f['properties']['qs:photos'], basestring):
+                f['properties']['qs:photos'] = str(f['properties']['qs:photos'])
```
Comment (nvkelso) [inline scripts/wof-fix-props]: Int

File: scripts/wof-fix-props [hunk 4]
```diff
@@ -0,0 +1,561 @@
...
+                fixed = True
+
+        if 'qs:photos_sr' in f['properties']:
+            if not isinstance(f['properties']['qs:photos_sr'], basestring):
+                f['properties']['qs:photos_sr'] = str(f['properties']['qs:photos_sr'])
```
Comment (nvkelso) [inline scripts/wof-fix-props]: Int

File: scripts/wof-fix-props [hunk 5]
```diff
@@ -0,0 +1,561 @@
...
+                fixed = True
+
+        if 'meso:local_id' in f['properties']:
+            if not isinstance(f['properties']['meso:local_id'], basestring):
+                f['properties']['meso:local_id'] = str(f['properties']['meso:local_id'])
+                fixed = True
+
+        if 'meso:pop' in f['properties']:
+            if not isinstance(f['properties']['meso:pop'], basestring):
+                f['properties']['meso:pop'] = str(f['properties']['meso:pop'])
```
Comment (nvkelso) [inline scripts/wof-fix-props]: Int

File: scripts/wof-fix-props [hunk 6]
```diff
@@ -0,0 +1,561 @@
...
+
+        # Fixup ne: properties
+        if 'ne:fips_10_' in f['properties']:
+            if not isinstance(f['properties']['ne:fips_10_'], basestring):
+                f['properties']['ne:fips_10_'] = str(f['properties']['ne:fips_10_'])
+                fixed = True
+
+        if 'ne:gdp_md_est' in f['properties']:
+            if not isinstance(f['properties']['ne:gdp_md_est'], basestring):
+                f['properties']['ne:gdp_md_est'] = str(f['properties']['ne:gdp_md_est'])
```
Comment (nvkelso) [inline scripts/wof-fix-props]: Int

File: scripts/wof-fix-props [hunk 7]
```diff
@@ -0,0 +1,561 @@
...
+                fixed = True
+
+        if 'ne:gdp_year' in f['properties']:
+            if not isinstance(f['properties']['ne:gdp_year'], basestring):
+                f['properties']['ne:gdp_year'] = str(f['properties']['ne:gdp_year'])
```
Comment (nvkelso) [inline scripts/wof-fix-props]: Int

File: scripts/wof-fix-props [hunk 8]
```diff
@@ -0,0 +1,561 @@
...
+                fixed = True
+
+        if 'ne:wikipedia' in f['properties']:
+            if not isinstance(f['properties']['ne:wikipedia'], basestring):
+                f['properties']['ne:wikipedia'] = str(f['properties']['ne:wikipedia'])
+                fixed = True
+
+        if 'ne:woe_id' in f['properties']:
+            if not isinstance(f['properties']['ne:woe_id'], basestring):
+                f['properties']['ne:woe_id'] = str(f['properties']['ne:woe_id'])
```
Comment (nvkelso) [inline scripts/wof-fix-props]: Int

File: scripts/wof-fix-props [hunk 9]
```diff
@@ -0,0 +1,561 @@
...
+                fixed = True
+
+        if 'ne:woe_id_eh' in f['properties']:
+            if not isinstance(f['properties']['ne:woe_id_eh'], basestring):
+                f['properties']['ne:woe_id_eh'] = str(f['properties']['ne:woe_id_eh'])
```
Comment (nvkelso) [inline scripts/wof-fix-props]: Int

File: scripts/wof-fix-props [hunk 10]
```diff
@@ -0,0 +1,561 @@
...
+                fixed = True
+
+        if 'qs:photos_sr' in f['properties']:
+            if not isinstance(f['properties']['qs:photos_sr'], basestring):
+                f['properties']['qs:photos_sr'] = str(f['properties']['qs:photos_sr'])
+                fixed = True
+
+        if 'qs:photos_1k' in f['properties']:
+            if not isinstance(f['properties']['qs:photos_1k'], basestring):
+                f['properties']['qs:photos_1k'] = str(f['properties']['qs:photos_1k'])
```
Comment (nvkelso) [inline scripts/wof-fix-props]: Int

File: scripts/wof-fix-props [hunk 11]
```diff
@@ -0,0 +1,561 @@
...
+                fixed = True
+
+        if 'qs:photos_9k' in f['properties']:
+            if not isinstance(f['properties']['qs:photos_9k'], basestring):
+                f['properties']['qs:photos_9k'] = str(f['properties']['qs:photos_9k'])
```
Comment (nvkelso) [inline scripts/wof-fix-props]: Int

File: scripts/wof-fix-props [hunk 12]
```diff
@@ -0,0 +1,561 @@
...
+                fixed = True
+
+        if 'qs:photos_9r' in f['properties']:
+            if not isinstance(f['properties']['qs:photos_9r'], basestring):
+                f['properties']['qs:photos_9r'] = str(f['properties']['qs:photos_9r'])
```
Comment (nvkelso) [inline scripts/wof-fix-props]: Int

File: scripts/wof-fix-props [hunk 13]
```diff
@@ -0,0 +1,561 @@
...
+                fixed = True
+
+        if 'qs:photos_all' in f['properties']:
+            if not isinstance(f['properties']['qs:photos_all'], basestring):
+                f['properties']['qs:photos_all'] = str(f['properties']['qs:photos_all'])
```
Comment (nvkelso) [inline scripts/wof-fix-props]: Int

File: scripts/wof-fix-props [hunk 14]
```diff
@@ -0,0 +1,561 @@
...
+                fixed = True
+
+        if 'qs:gn_id' in f['properties']:
+            if not isinstance(f['properties']['qs:gn_id'], basestring):
+                f['properties']['qs:gn_id'] = str(f['properties']['qs:gn_id'])
```
Comment (nvkelso) [inline scripts/wof-fix-props]: Int

File: scripts/wof-fix-props [hunk 15]
```diff
@@ -0,0 +1,561 @@
...
+                fixed = True
+
+        if 'qs:gn_pop' in f['properties']:
+            if not isinstance(f['properties']['qs:gn_pop'], basestring):
+                f['properties']['qs:gn_pop'] = str(f['properties']['qs:gn_pop'])
```
Comment (nvkelso) [inline scripts/wof-fix-props]: Int

**Review by nvkelso — CHANGES_REQUESTED** [review]
Nice work, thank you!

~15 of these should be int (not string).

**Comment by vicchi** [comment]
@nvkelso Let me check these tomorrow morning; probably a copy-and-paste snarl up on my part, but I'll double check

File: VERSION [hunk 16]
```diff
@@ -1 +1 @@
-0.063
```
Comment (thisisaaronland) [inline VERSION:1]: We've been gradually migrating all the version numbers to use semver syntax so this is a good opportunity to bump this to ... `0.1.0` which doesn't really mean much beyond "bigger than "0.063"

File: scripts/wof-find-property [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+        elif options.property in f['properties']:
+            p, filename = os.path.split(path)
+            value = f['properties'][options.property]
+            if isinstance(value, str) or isinstance(value, unicode):
+                value = value.encode('utf-8')
+
+            print('{} {} - {} ({})'.format(filename, options.property, value, type(value)))
+
+    except:
+        print('\nFatal exception caught in {}'.format(path))
```
Comment (thisisaaronland) [inline scripts/wof-find-property:37]: For warnings and errors, we should use the `logging` facilities.

File: VERSION (hunk 16 above)
Comment (vicchi) [reply VERSION:1]: `0.064` begone; it's all semver from now on. Also fixed in commit `992be82`

File: scripts/wof-fix-props (hunk 1 above)
Comment (vicchi) [reply scripts/wof-fix-props]: Fixed in commit `5bcad4c`

File: scripts/wof-fix-props (hunk 2 above)
Comment (vicchi) [reply scripts/wof-fix-props]: Fixed in commit `5bcad4c`

File: scripts/wof-fix-props (hunk 3 above)
Comment (vicchi) [reply scripts/wof-fix-props]: Fixed in commit `5bcad4c`

File: scripts/wof-fix-props (hunk 4 above)
Comment (vicchi) [reply scripts/wof-fix-props]: Fixed in commit `5bcad4c`

File: scripts/wof-fix-props (hunk 5 above)
Comment (vicchi) [reply scripts/wof-fix-props]: Fixed in commit `5bcad4c`

File: scripts/wof-fix-props (hunk 6 above)
Comment (vicchi) [reply scripts/wof-fix-props]: Fixed in commit `5bcad4c`

File: scripts/wof-fix-props (hunk 7 above)
Comment (vicchi) [reply scripts/wof-fix-props]: Fixed in commit `5bcad4c`

File: scripts/wof-fix-props (hunk 8 above)
Comment (vicchi) [reply scripts/wof-fix-props]: Fixed in commit `5bcad4c`

File: scripts/wof-fix-props (hunk 9 above)
Comment (vicchi) [reply scripts/wof-fix-props]: Fixed in commit `5bcad4c`

File: scripts/wof-fix-props (hunk 10 above)
Comment (vicchi) [reply scripts/wof-fix-props]: Fixed in commit `5bcad4c`

File: scripts/wof-fix-props (hunk 11 above)
Comment (vicchi) [reply scripts/wof-fix-props]: Fixed in commit `5bcad4c`

File: scripts/wof-fix-props (hunk 12 above)
Comment (vicchi) [reply scripts/wof-fix-props]: Fixed in commit `5bcad4c`

File: scripts/wof-fix-props (hunk 13 above)
Comment (vicchi) [reply scripts/wof-fix-props]: Fixed in commit `5bcad4c`

File: scripts/wof-fix-props (hunk 14 above)
Comment (vicchi) [reply scripts/wof-fix-props]: Fixed in commit `5bcad4c`

File: scripts/wof-fix-props (hunk 15 above)
Comment (vicchi) [reply scripts/wof-fix-props]: Fixed in commit `5bcad4c`

File: scripts/wof-find-property (hunk 17 above)
Comment (vicchi) [reply scripts/wof-find-property:37]: Agreed; fixed in commit `6590205`

**Review by nvkelso — APPROVED** [review]
LGTM, thanks!

**Comment by vicchi** [comment]
@nvkelso @thisisaaronland Thanks! Happy for me to merge this?

**Pull request merged**